src/main/java/hudson/security/ProjectMatrixAuthorizationStrategy.java

```diff
@@ -56,16 +58,29 @@ public ACL getACL(Job<?,?> project) {
             SidACL projectAcl = amp.getACL();
 
             if (!amp.isBlocksInheritance()) {
-                projectAcl = projectAcl.newInheritingACL(getACL(project.getParent()));
+                final ACL parentAcl = getACL(project.getParent());
+                return inheritingACL(parentAcl, projectAcl);
+            } else {
+                return projectAcl;
             }
-
-            return projectAcl;
         } else {
             return getACL(project.getParent());
         }
     }
 
-    public SidACL getACL(ItemGroup g) {
```

Comment:
Unfortunately this is incompatible for `cloudbees-folder`:

```
java.lang.NoSuchMethodError: hudson.security.ProjectMatrixAuthorizationStrategy.getACL(Lhudson/model/ItemGroup;)Lhudson/security/SidACL;
	at com.cloudbees.hudson.plugins.folder.AbstractFolder.getACL(AbstractFolder.java:530)
```

We need to invert the dependency here.